bridges/snowbridge/primitives/router/src/inbound/mod.rs

```diff
@@ -291,6 +381,100 @@ where
 			[GlobalConsensus(network), AccountKey20 { network: None, key: token.into() }],
 		)
 	}
+
+	/// Constructs an XCM message destined for AssetHub that withdraws assets from the sovereign
+	/// account of the Gateway contract and either deposits those assets into a recipient account or
+	/// forwards the assets to another parachain.
+	fn convert_send_native_token(
+		message_id: H256,
+		chain_id: u64,
+		token_id: TokenId,
+		destination: Destination,
+		amount: u128,
+		asset_hub_fee: u128,
+	) -> Result<(Xcm<()>, Balance), ConvertMessageError> {
+		let network = Ethereum { chain_id };
+		let asset_hub_fee_asset: Asset = (Location::parent(), asset_hub_fee / 2).into();
+
+		let (dest_para_id, beneficiary, dest_para_fee) = match destination {
+			// Final destination is a 32-byte account on AssetHub
+			Destination::AccountId32 { id } =>
+				(None, Location::new(0, [AccountId32 { network: None, id }]), 0),
+			// Final destination is a 32-byte account on a sibling of AssetHub
+			Destination::ForeignAccountId32 { para_id, id, fee } =>
+				(Some(para_id), Location::new(0, [AccountId32 { network: None, id }]), fee),
+			// Final destination is a 20-byte account on a sibling of AssetHub
+			Destination::ForeignAccountId20 { para_id, id, fee } =>
+				(Some(para_id), Location::new(0, [AccountKey20 { network: None, key: id }]), fee),
+		};
+
+		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
+		let total_fee_asset: Asset = (Location::parent(), total_fees).into();
+
+		let asset_loc =
+			ConvertAssetId::convert(&token_id).ok_or(ConvertMessageError::InvalidToken)?;
+
+		let mut reanchored_asset_loc = asset_loc.clone();
+		reanchored_asset_loc
+			.reanchor(&GlobalAssetHubLocation::get(), &UniversalLocation::get())
+			.map_err(|_| ConvertMessageError::CannotReanchor)?;
+
+		let asset: Asset = (reanchored_asset_loc, amount).into();
+
+		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+
+		let mut instructions = vec![
+			ReceiveTeleportedAsset(total_fee_asset.clone().into()),
+			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
+			DescendOrigin(PalletInstance(inbound_queue_pallet_index).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			WithdrawAsset(asset.clone().into()),
+		];
+
+		match dest_para_id {
+			Some(dest_para_id) => {
+				let dest_para_fee_asset: Asset = (Location::parent(), dest_para_fee).into();
+
+				instructions.extend(vec![
+					// `SetAppendix` ensures that `fees` are not trapped in any case
+					SetAppendix(Xcm(vec![DepositAsset {
+						assets: AllCounted(1).into(),
+						beneficiary: Location::new(1, [Parachain(dest_para_id)]),
+					}])),
+					// Perform a reserve withdraw to send to destination chain. Leave half of the
+					// asset_hub_fee for the delivery cost
+					InitiateReserveWithdraw {
```

Comment:
This will only work for assets where the reserve chain is the destination parachain. Assets where the reserve chain is asset hub will result in funds being lost. For example, DOT will not be held in reserve by Asset Hubs sovereign on the destination chain, so the withdraw from reserve will fail on the destination and be burnt on Asset Hub. The UX layer can stop this kind of transfer if we choose not to rectify this.

Reply:
Challege is that destination chain won't allow asset if it does not trust AH as reserve chain. Or the trust model between Parachain and AH can even be teleport based. 

Consider all the cases it's complicated to build the xcm in BH support everything. I'd prefer to make things easy start with a limited scope at the begining.  

Yeah, we can check from UX layer to stop transfer not support.

Reply:
This comment can be resolved @yrong

Reply:
I want to reopen this conversation, this looks incorrect to me. Which assets would this actually support like this?

The long-term plan is to have parachain native assets registered as foreign assets on AH, which means they would be teleportable between para and AH. So these won't work like this.

AH native assets like USDT, USDC or bridged assets like KSM (and ecosystem) also won't work as those have reserve on local AH.

The solution here doesn't lie on-chain, but it lies offchain (see layer 2 from [XCM Asset Transfer Tooling section](https://github.com/paritytech/polkadot-sdk/issues/4736)), aka in your UI.
Type of transfer between AH and final parachain needs to come in from UI as a param (InitiateReserveWithdraw, DepositReserveAsset or InitiateTeleport). The UI does offchain checks to on-chain configs to find the right one.

Reply:
> Which assets would this actually support like this?

Multi-hop is only for native token on 3rd Parachain(e.g. GLMR for Moonbeam).

Reply:
> trust model between Parachain and AH is reserved based(e.g. GLMR for Moonbeam)

At the ecosystem level we are pushing to change this to be teleport (e.g. register GLMR as foreign asset on AH). Some Parachains will register their native assets sooner, some later. But 100% the bridge transfer will brake for multiple parachains' native tokens if you hardcode to always use `InitiateReserveWithdraw`.

Reply:
Yeah, that makes sense.

Can we just limit the scope of this PR and improve it later? As first step we can disable the multi-hop support easily on Ethereum side.

Reply:
Yes we can limit the scope now and improve it later.

Reply:
For limiting scope now, it's best to simply remove this branch - and only support AH as final destination.

The current code is actually incorrect and can lead to issuance problems because it does `InitiateReserveWithdraw` for DOT too, which incorrectly burns DOT on AH and on destination parachain it withdraws it from AH's sovereign account.

I will open an issue/PR for explicitly disabling/erroring in the `xcm-executor` on trying to `InitiateReserveWithdraw` for an unsupported `(asset, reserve)` combination.

Reply:
Cool! Though we do have the use case in use by MYTH https://app.snowbridge.network/history#0x71a6e41c7e60139af298af858a5ad44389e4819bf72e227bde9b62fbf281a04a
So I'd prefer to leave the branch here for limited support and we do the check on solidity or UI side.

Reply:
I am currently making a PR to remove this branch and disable the destination transfer feature for PNA tokens only.

Reply:
yes, disabled only for PNA tokens.

Ethereum tokens can be hopped through AH to another para because their reserve location in the Polkadot Ecosystem is always AH.

Once you remove this incorrect code for PNAs, we can merge the PR from my pov.

Reply:
Addressed in the following PRs:
SDK: https://github.com/Snowfork/polkadot-sdk/pull/179
Snowbridge: https://github.com/Snowfork/snowbridge/pull/1284

Reply:
Some cleanup in https://github.com/paritytech/polkadot-sdk/pull/5546/commits/43e546b176835d6b023924a25b4762b252d488db